[13.0][MIG] l10n_es_aeat_mod303

ValentinVinagre opened the pull request:
En proceso de migración, falta:
- [x] Casos de uso.
- [x] Adaptación de código en detalle.
- [x] Adaptación de Tests.

Depende de:
- [x] https://github.com/OCA/l10n-spain/pull/1298


El PR: https://github.com/OCA/l10n-spain/pull/1340/files.

Commit:
[IMP] l10n_es_aeat: Algo de PEP8 y refactorización de código
[MIG] l10n_es_aeat_mod303: Migración a v7 del módulo con las siguientes mejoras:
- Convenciones de la comunidad.
- PEP8 en la mayoría de código.
- Icono para el módulo.
- Adaptación de vistas a las posibilidades de la v7.
- Restringidas las cuentas bancarias a las de la compañía.

Commit:
Launchpad automatic translations update.

Commit:
[IMP] l10n_es_aeat: Mejorado mecanismo de herencia para reaprovechar más código.
[IMP] l10n_es_aeat: Añadido wizard para ver el resultado de la exportación sin necesidad de tener el módulo document.
[IMP] l10n_es_aeat: Puesta transición de estados sin workflow.
[IMP] l10n_es_aeat_mod303: Adaptado al nuevo mecanismo. Se elimina workflow anterior mediante un script de migración.

Commit:
Launchpad automatic translations update.

acysos committed:
[FIX] l10n_es_aeat: compatibilidad de la función export_to_boe para el modelo 340

Commit:
Launchpad automatic translations update.

Commit:
[FIX] l10n_es_aeat_mod130/mod303: import corregido para que valga también para la v8.

Commit:
Launchpad automatic translations update.

pedrobaeza committed:
[MIG] Movidos todos los módulos a __unported__
[MIG] Eliminados módulos no migrados a v7

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Modelo 303 de la AEAT totalmente refactorizado

Presentación del Modelo AEAT 303
================================

* Incluye un motor genérico para obtener los totales de códigos de impuestos
  de manera parametrizada y con posibilidad de versionado por fechas
* Incluye un motor de parametrización de las exportaciones a formato BOE
  con posibilidad de versionado por fechas.
* Se incluyen las parametrizaciones de ambos motores para el 303.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Avoid errors in first initialization

oca-transbot committed:
OCA Transbot updated translations from Transifex

juanalfonsopr committed:
[FIX] l10n_es_aeat_mod303: Corregido cálculo de campo resultado_liquidacion y corregido formulario de mod303, sustituido campo casilla_46 repetido por campo casilla_69.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Añadir documentos explicativos del modelo

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Enfoque definitivo

======================
AEAT - Prorrata de IVA
======================

Módulo para gestionar la prorrata del IVA en las declaraciones de la AEAT (por
el momento, en el modelo 303), según el artículo 92 de la Ley 37/1992, de 28 de
diciembre, del Impuesto sobre el Valor Añadido.

Uso
===

Realizando el modelo 303, se puede seleccionar si se desea aplicar prorrata de
IVA.

Prorrata general
----------------

En el caso de seleccionar "Prorrata general", aparece un nuevo campo llamado
"Porcentaje de prorrata de IVA" que puede ser rellenado a mano.

Además, cuando se realice el asiento de regularización, la parte proporcional
de la cuota no deducida se compensará como gasto.

Se ha incluido un calculador del porcentaje de prorrata, utilizable en 2 casos:

* Al comienzo de un nuevo ejercicio, para determinar el porcentaje provisional
  a aplicar hasta el último periodo de ese ejercicio.
* En la última declaración del ejercicio, para determinar el porcentaje
  definitivo de prorrata, y sobre el que se compensará la diferencia.

Problemas conocidos / Hoja de ruta
==================================

* La prorrata especial de IVA no está contemplada aún.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Algunos arreglos visuales

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat/l10n_es_aeat_mod303: Crear asiento de regularización cuando se hace la declaración

* Generador del asiento de regularización con cargo a un proveedor "Agencia
  Estatal de Administración Tributaria" creado al efecto.

Configuración
=============

Para activar la creación del asiento de regularización en un modelo, hay que
poner en el modelo correspondiente el campo allow_posting a True, y establecer
en la configuración de impuestos los conceptos que se regularizarán con el
flag "to_regularize".

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Don't ask for the bank account

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[ADD] l10n_es_aeat_mot303, casilla 77

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Mejor herencia de impuestos

El anterior método utilizado, al tener una clase y un many2one genérico,
no permitía poner los datos en múltiples modelos hijos, ya que se repetían
los IDs, y cogía los datos de un modelo en otros.

Con este nuevo sistema, se sigue utilizando una tabla, pero con doble
índice: el modelo y el ID, parecido a un campo reference, pero gestionado
por código propio en el método de cálculo.

El 303 está adaptado a este nuevo sistema, y se incluye un script de
migración que conserve los datos.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Evitar problemas en campos related no readonly. Fixes #290

Cuando se tienen campos related, aunque el valor no se cambie,
Odoo lanza una escritura en el objeto relacionado cada vez que se
guarda el objeto principal. En el caso de la moneda asociada a la
compañía, hace que aquellos usuarios que no tengan permiso para
modificar la compañía (la mayoría), falle en la creación del modelo.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Export fields [27] and [45]

oca-transbot committed:
OCA Transbot updated translations from Transifex

omar7r committed:
[IMP] l10n_es_aeat_mod303: Se añade la gestión del iva diferido en el 303

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Include analytics + tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

antespi committed:
[MIG] l10n_es_aeat/303: Rest of the migration

pedrobaeza committed:
[IMP] l10n_es_aeat: Several warnings and imps

pedrobaeza committed:
[MIG] l10n_es_aeat/mod303: Migration to 10.0

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Multi-company security rule

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt new BOE 2017 format

pedrorgil committed:
[FIX] aeat_mod_303. Corrección tipos de exportación (#465)

fuentes73 committed:
[FIX] l10n_es_aeat_mod303: Añadir la cuenta bancaria en la exportación

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Show compensate flag

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Record design link

pedrobaeza committed:
[FIX] l10n_es_aeat: Avoid multi-company error due to depends

Debido a que se utiliza una sola tabla para almacenar las líneas
de los impuestos de todos los modelos, hay una limitación en el ORM de Odoo
cuando se coloca el campo one2many de dichas líneas (tax_line_ids) como
dependencia en la definición del cálculo de un campo (entrada con
@api.depends), que recalcula los campos calculados de todos los modelos con el
mismo ID que el del registro en curso, lo que puede ser un problema en entornos
multi-compañía. Una solución a ello (aunque no evita el recálculo), es poner en
esos campos calculados `compute_sudo=True`.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Conditional expression is reversed + 303 with conditional

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Change end year in previous export format

gurneyalex committed:
[REF] i18n: Updating translation terms from weblate [ci skip]

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10_es_aeat_mod303: Import 2017 data file

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Set date_end on 2014 export config

oihane committed:
[MIG] l10n_es_aeat_mod303: Migration to 11.0

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Automatic compensation and return improvement

Si no estás inscrito en el REDEME (Registro de Devolución Mensual), y la declaración
sale negativa, la única posibilidad es la de compensar, por lo que no tiene sentido
que se tenga que marcar manualmente la casilla, ya que si olvida y se exporta, Hacienda
no lo admite.

Por eso, este cambio detecta automáticamente esta situación y coloca el tipo de resultado,
eliminando la casilla extra.

Se incluye script de migración para restituir correctamente los datos de anteriores
declaraciones.

También se hace obligatoria la cuenta bancaria cuando la declaración sea a devolver, ya
que es también un dato obligatorio al presentarlo a Hacienda.

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Migration script

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Migration not needed

Due to compute non stored field

JordiBForgeFlow committed:
- Cambia cuenta de contrapartida en función del resultado.
- En a casilla de importes a compensar debe constar, EN POSITIVO, las cuotas a compensar procedentes de periodos anteriores.
- Incluye mensaje para advertir que existen cuotas a compensar en periodos anteriores.
- Borra campo 'bank_account_id', que está ya duplicado en 'partner_bank_id' e incluye script de migración.

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt to latest 2017 BOE format

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: BOE 2018 format

oca-transbot committed:
OCA Transbot updated translations from Transifex

luismontalba committed:
[IMP] l10n_es_aeat_mod303: Descuentos_cuota_IVA

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Improve tests for handling negative price units

carlosdauden committed:
[FIX] l10n_es_aeat_mod303: dict.iteritems() don't exist in python3. Change iteritems() to iter(items())

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod303: Fix refinements

* No need to change test structure with create
* Optimize code
* Fix tests in l10n_es_vat_prorrate
* Translations
* Other minor things

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Perform strict search of counterpart account

Odoo's operator "like" encloses the text to search into "%s", making
the search to match with accounts like "157000" and others as addition
to the expected "570000".

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Fix result when compensating

* Don't show the warning for draft reports
* When you have to return result, don't propagate the result to next report
* Put only once the warning

Abranes committed:
[FIX] l10n_es_aeat_mod303: Change export_type of Nombre to alphabetic

oca-travis committed:
[UPD] Update l10n_es_aeat_mod303.pot

mtbochaca committed:
Translated using Weblate (català)

Currently translated at 100,0% (45 of 45 strings)

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_aeat_mod303
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_aeat_mod303/ca/

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Include negative amounts in refund tax amounts

They were ignored previously

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Migration to 12.0

* Standard procedure
* Adapted tests
* README by fragments

oca-travis committed:
[UPD] Update l10n_es_aeat_mod303.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Add support for 2018 4T:

* BOE format
* Exonerated from 390

  * Fields
  * Mappings
  * BOE export

* README update
* Little change in AEAT engine
* Tests adapted + expanded

oca-travis committed:
[UPD] Update l10n_es_aeat_mod303.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Proper view when not exonerated

As numbers have changed, the view was not adjusted.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Don't inform 347 if not exonerated or not last period

Inform this field only in the case where this proceeds.

pedrobaeza committed:
[IMP] l10n_es_aeat: Add active field to export configs

This way, we don't initially see all the subconfigs for each model and we can concentrate
on main export configurations. This is got through `active` field default behaviour.
Archived config are still directly searchable through the subconfig selector in the
configuration line.

All models export configs have been adapted to this new approach.

pedrobaeza committed:
[IMP] l10n_es_aeat: Add chatter + activities to AEAT models

* Include chatter (mail.thread)
* Include activities (mail.activity.mixin)
* Track AEAT report state change
* Include new admin user in AEAT group by default

CarlesAntoli committed:
Translated using Weblate (Spanish)

Currently translated at 32.3% (54 of 167 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod303
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod303/es/

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Include both signs (debit and credit) in tax fee fields

Given this regular invoice:

1, IVA 10%, 10 €, 1 € tax
2, IVA 21%, -1 €, -0,21 € tax

that can happen when you return some packages (like returnable bottles), you have
a valid invoice with total positive amount to due, but in 303, the -0.21 € of the
tax fee is not included as there's no other positive tax fee to compensate that
negative amount, and the tax map line indicated that only credit operations are
included.

As the move type is already checked for avoiding to include refunds, we can change
to include both move lines types (debit and credit) for this. In fact, in the
base, both are already included.

As this can happen as well with other taxes and with supplier invoices, all
involved fields are changed in the same commit.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Add field 84

It's the sum of all the year of field 61, so the definition is the same.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt test to previous change

Field 88 doesn't include the new field 84

MiquelRForgeFlow committed:
[UPD] Remove deprecated from_string()

pedrobaeza committed:
[IMP+FIX] l10n_es_aeat: Move get taxes method to company + cleaning in tests

This way, we can reuse it in other modules that don't inherit AEAT report like SII.

We also clean a non needed reading of the tax description field.

Tests were still using description method for populating data, so we switch to new
system and perform needed changes in dependent tests.

s

ValentinVinagre commented:
Hola @pedrobaeza @MiquelRForgeFlow ,
estoy realizando los test del 303 pero cuando estoy realizando la comprobación de los campos total_devengado, total_deducir....
por lo visto tiene un problema con la cache y no lo guarda ¿sabes de alguna idea para solucionarlo?
he estado probando con el flush y el invalidate_cache pero nada....

Gracias

pedrobaeza commented:
Buenas, ahora mismo no sabría decirte. Tengo que revisar estos PRs a ver.

MiquelRForgeFlow commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -15,323 +18,490 @@ class TestL10nEsAeatMod303Base(TestL10nEsAeatModBase):
     }
 
     def setUp(self):
         super(TestL10nEsAeatMod303Base, self).setUp()
```
```suggestion
    @classmethod
    def setUpClass(cls):
        super().setUpClass()
```
recuerda que ahora el test es savepointcase

MiquelRForgeFlow commented on l10n_es_aeat_mod303/security/l10n_es_aeat_mod303_security.xml:
```diff
@@ -0,0 +1,13 @@
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >[('company_id', 'child_of', [user.company_id.id])]</field>
```
```suggestion
        >[('company_id', 'in', company_ids)]</field>
```

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Avoid error importing BOE due to developer VAT

AEAT now forces to be a proper VAT format. We adjust at the same time the version
and some old year references.

acysos committed:
[FIX] l10n_es_aeat_mod303: fix 303 map for field number 60

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod303: Fix tests for fields 60/61 + improve ROADMAP

HaraldPanten commented:
Hola @ValentinVinagre 

Has tenido en cuenta https://github.com/OCA/l10n-spain/pull/1340 y https://github.com/OCA/l10n-spain/pull/1262 ?

Saludos,

ValentinVinagre commented:
Hola @HaraldPanten , los tendré en cuenta cuando continúe .

angelmoya committed:
[FIX] 303,facturas rectivicativas en 93 y 94

rlizana committed:
[12.0] l10n_es_aeat_*: [FIX] Caracteres no válidos 'NIF Empresa Desarrollo'

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 32.9% (55 of 167 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod303
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod303/es/

acysos commented:
@ValentinVinagre No veo el commit de black, isort https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0

ValentinVinagre commented:
> @ValentinVinagre No veo el commit de black, isort https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0

Me lo habré comido.... ahora lo arreglo disculpa.

MiquelRForgeFlow commented:
Por cierto, se añadieron nuevos commits en v12 https://github.com/OCA/l10n-spain/commits/12.0/l10n_es_aeat_mod303, que tocaría añadir tambien aquí 🤔

Y esto tambien va para los otros PR

ValentinVinagre commented:
> Por cierto, se añadieron nuevos commits en v12 https://github.com/OCA/l10n-spain/commits/12.0/l10n_es_aeat_mod303, que tocaría añadir tambien aquí 🤔
> 
> Y esto tambien va para los otros PR

Justo ahora he visto eso y lo estaba revisando.

ValentinVinagre commented:
El PR: https://github.com/OCA/l10n-spain/pull/1340/files.
No se puede incluir en la migración debido a que Odoo v.13 no tiene el impuesto "account_tax_template_s_iva_ns_b" https://github.com/odoo/odoo/pull/34044 creado por lo que va a realizar otro PR con ese IMP para cuando Odoo haga el Merge poder incluirlo y que no sea un stop.
Se acaba de incluir el PR: https://github.com/OCA/l10n-spain/pull/1262/files

ValentinVinagre committed:
[IMP] l10n_es_aeat_mod303: black, isort, prettier

ValentinVinagre committed:
[MIG] l10n_es_aeat_mod303: Migration to 13.0

ValentinVinagre commented:
@MiquelRForgeFlow creo que ya me he agregado todo el historial actual de commits.

MiquelRForgeFlow commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -440,19 +445,21 @@ def setUp(self):
         )
 
 
+@common.tagged("post_install", "-at_install")
```
```suggestion
```
al final quítalo, porque tambien se quitó eso en el TestL10nEsAeatModBase del l10n_es_aeat

MiquelRForgeFlow commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -12,6 +13,7 @@
 _logger = logging.getLogger("aeat.303")
 
 
+@common.tagged("post_install", "-at_install")
```
```suggestion
```
al final quítalo, porque tambien se quitó eso en el TestL10nEsAeatModBase del l10n_es_aeat

MiquelRForgeFlow reviewed (approved):
LGTM

HaraldPanten reviewed (approved):
Functional review 👍

acysos reviewed (approved):
Funciona correcta. Pruebas funcionales.

pedrobaeza commented:
Lo del IVA no sujeto va a estar fusionado hoy en Odoo y esta noche en OCB, así que podemos esperar a fusionarlo para tener todo junto.

ValentinVinagre commented:
> Lo del IVA no sujeto va a estar fusionado hoy en Odoo y esta noche en OCB, así que podemos esperar a fusionarlo para tener todo junto.

Entonces voy a pasar los cambios de los commits pertinentes. Gracias por el aviso.

ValentinVinagre commented:
Se han incluido los commits sobre IVA no sujeto, por lo que hasta que no esté el impuesto mergeado en odoo dará error el travis.

ValentinVinagre commented:
All green 👍

MiquelRForgeFlow commented:
añade script de migracion, como hice en https://github.com/OCA/l10n-spain/pull/1307.

ValentinVinagre commented:
> añade script de migracion, como hice en #1307.

¿Debería de añadir en el "requirements.txt" también el "openupgradelib" o como lo has añadido en el 111 ya es suficiente?

MiquelRForgeFlow commented on l10n_es_aeat_mod303/migrations/13.0.1.0.0/pre-migration.py:
```diff
@@ -0,0 +1,17 @@
+# Copyright 2020 ForgeFlow
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from openupgradelib import openupgrade
```
```suggestion
from openupgradelib import openupgrade  # pylint: disable=W7936
```

ValentinVinagre replied:
done

pedrobaeza commented on l10n_es_aeat_mod303/README.rst:
```diff
@@ -60,25 +60,14 @@ Known issues / Roadmap
   `l10n_es_aeat_mod303_cash_basis`.
 * La prorrata del IVA está contemplada por el módulo adicional
   `l10n_es_aeat_vat_prorrate`.
-* Existen 2 casos de IVA no sujeto que van a la casilla 61 del modelo, que no
```
Por qué has eliminado esto?

ValentinVinagre replied:
Si no me equivoco, lo quité cuando separé en dos pr's, ahora está indicado.

pedrobaeza commented on l10n_es_aeat_mod303/models/mod303.py:
```diff
@@ -24,10 +25,7 @@ class L10nEsAeatMod303Report(models.Model):
-                ("code", "like", "4750%"),
-                ("company_id", "=", self._default_company_id().id),
-            ]
+            [("code", "like", "4750%"), ("company_id", "=", self.company_id.id)]
```
Esto no vale porque la compañía aún no está establecida. Puedes hacer este campo un calculado escribible en su lugar.

ValentinVinagre replied:
No acabo de entender el concepto.
En un principio el campo "company_id" se crea en el modelo "l10n.es.aeat.report" el cual coge un valor por defecto, por lo tanto debería de funcionar.
En caso de que no lo recogiera correctamente podría poner: self.company_id or self.env.company, para que lo recoja del entorno.

ValentinVinagre replied:
@pedrobaeza

pedrobaeza replied:
Como digo, aquí el por defecto no está aún establecido. Haz el campo computed escribible como te comento que es la solución más sencilla.

ValentinVinagre replied:
¿Te refieres al campo "counterpart_account_id"? si no, no acabo de entenderte.

pedrobaeza replied:
Eso es:
```python
    counterpart_account_id = fields.Many2one(
        comodel_name="account.account",
        string="Counterpart account",
        compute="_compute_counterpart_account_id",
        store=True,
        readonly=False,
        domain="[('company_id', '=', company_id)]",
    )

    @api.depends("company_id", "result_type")
    def _compute_counterpart_account_id(self):
        for record in self:
            code = "%s%%" % _ACCOUNT_PATTERN_MAP.get(record.result_type, "4750"),
            record.counterpart_account_id = self.env["account.account"].search(
                [("code", "like", code), ("company_id", "=", self.company_id.id)]
            )
```
y elimina el código de `calculate` que hace lo mismo.

ValentinVinagre replied:
Hecho.

pedrobaeza commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -574,21 +576,22 @@ def test_model_303(self):
         # Check change of period type
         self.model303_4t.period_type = "1T"
-        self.model303_4t.onchange_period_type()
+        self.model303_4t._compute_dates()
```
Esto ya no debería ser necesario invocarlo directamente

ValentinVinagre replied:
Si lo elimino no realiza la función correctamente (entiendo que este mismo caso también pasaba en las anteriores versiones).

pedrobaeza replied:
Ya, pero esto siempre esconde algo detrás, y en este caso ha destapado esto:

https://github.com/OCA/l10n-spain/pull/1328/files?file-filters%5B%5D=.py#diff-de75969fd945555d566bd15ddc01d3fcR344-R346

Se estaban asignando unos campos en este compute, y aunque eso funcionaba antes, ahora hay que regularizar esto:

- Quitando esa sobreescritura del compute
- Haciendo esos 2 campos calculados escribibles, y poniendo para cada uno correctamente sus depends y su código de cálculo.

ValentinVinagre replied:
Hecho.

pedrobaeza commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -574,21 +576,22 @@ def test_model_303(self):
-    def test_model_303_negative_special_case(self):
-        self.taxes_sale = {
+    @classmethod
+    def change_taxes_negative_special_case(cls):
```
Por qué ponerlo en un método aparte? Se está reutilizando en algún otro sitio?

ValentinVinagre replied:
Debido a que taxes_sale, taxes_purchase, taxes_result son variables de clase, no permite cambiarlos si no es en una función @classmethod. Entiendo que es por los nuevos tipos de test.

pedrobaeza replied:
Esto no es cierto, ya que las variables sí que se pueden cambiar en cualquier sitio. Solo cambia si se referencian como `cls.<algo>` o `self.<algo>`

ValentinVinagre replied:
Según he estado viendo, diría que es correcto realizarlo de esta manera, te pongo un ejemplo:
La variable "accounts" es una variable de clase que se define en el siguiente test.
https://github.com/OCA/l10n-spain/blob/67d71a55d1080f0f270446e022fdd68c047d892f/l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py#L12

Para poder realizar un cambio en ella, lo debe de realizar con una función con el decorador `@classmethod`, como lo hace a continuación:
https://github.com/OCA/l10n-spain/blob/67d71a55d1080f0f270446e022fdd68c047d892f/l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py#L52-L53

Es el mismo caso que este caso, las variables `taxes_sale`,  `taxes_purchase`,  `taxes_result`, son variables definidas en la clase y si no las cambio de esta manera no surte efecto:
https://github.com/OCA/l10n-spain/blob/34dff17e87ec277a113ffcb001c5e0aab57be79f/l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py#L12-L16

pedrobaeza commented on l10n_es_aeat_mod303/data/aeat_export_mod303_2017_data.xml:
```diff
@@ -1208,7 +1208,7 @@
         <field name="export_config_id" ref="aeat_mod303_2017_sub02_export_config" />
         <field name="name">Indicador de página complementaria</field>
-        <field name="expression">${object.type == 'C'}</field>
+        <field name="expression">${object.result_type == 'C'}</field>
```
No debería ser `statement_type` en todos estos?

ValentinVinagre replied:
Corrrecto,  he cambiado los que estaban por "type" a "statement_type" por el cambio de campo, tanto en el 2017 como en el 2018. El resto si no me equivoco son correctos.

ValentinVinagre commented:
Hola @pedrobaeza , @MiquelRForgeFlow ,
el travis da un error de permisos cuando realiza los test del 303, por lo visto lo ejecuta un usuario con uid: 11, que corresponde al creado en los test de la "l10n_es_aeat"
https://github.com/OCA/l10n-spain/blob/67d71a55d1080f0f270446e022fdd68c047d892f/l10n_es_aeat/tests/test_l10n_es_aeat_mod_base.py#L291

No entiendo porque coge por defecto ese usuario para ejecutar los test, no sé si le tengo que especificar que lo haga todo con el admin o con el "account_manager" o es cosa del "l10n_es_aeat".
En local me funcionan bien los test, pero en el travis no.

pedrobaeza commented:
Poner `with_user` a todo no debe ser la solución. Si escribes pidiendo ayuda, espérate y deja el estado en el que pediste la ayuda para que así podamos verlo y diagnosticarlo. Ahora que has tocado, ya no se puede ver.

ValentinVinagre commented:
> Poner `with_user` a todo no debe ser la solución. Si escribes pidiendo ayuda, espérate y deja el estado en el que pediste la ayuda para que así podamos verlo y diagnosticarlo. Ahora que has tocado, ya no se puede ver.

Tienes toda la razón, disculpa, lo he dejado como estaba. @pedrobaeza ya se puede ver el error que ayer el travis no se lanzaba por la tarde.

HaraldPanten commented:
Hola @MiquelRForgeFlow @pedrobaeza alguna idea sobre cómo solventar el error de Travis? da un error de permisos pero no tiene mucho sentido...

pedrobaeza commented:
Si sale verde, se puede fusionar

/ocabot merge nobump
